pageserver/src/layered_repository.rs

```diff
@@ -230,16 +230,37 @@ impl Repository for LayeredRepository {
 
         let timelines = self.timelines.lock().unwrap();
         for (timelineid, timeline) in timelines.iter() {
-            walreceiver::stop_wal_receiver(*timelineid);
-            // Wait for syncing data to disk
-            trace!("repo shutdown. checkpoint timeline {}", timelineid);
-            timeline.checkpoint(CheckpointConfig::Forced)?;
-
-            //TODO Wait for walredo process to shutdown too
+            shutdown_timeline(*timelineid, timeline.as_ref())?;
         }
 
         Ok(())
     }
+
+    fn unload_timeline(&self, timeline_id: ZTimelineId) -> Result<()> {
+        let mut timelines = self.timelines.lock().unwrap();
+        let removed_timeline = match timelines.remove(&timeline_id) {
+            Some(timeline) => timeline,
+            None => {
+                warn!("Timeline {} not found, nothing to remove", timeline_id);
+                return Ok(());
+            }
+        };
+        drop(timelines);
+        shutdown_timeline(timeline_id, removed_timeline.as_ref())?;
+
+        Ok(())
+    }
+}
+
+fn shutdown_timeline(
+    timelineid: ZTimelineId,
+    timeline: &LayeredTimeline,
+) -> Result<(), anyhow::Error> {
+    walreceiver::stop_wal_receiver(timelineid);
+    trace!("repo shutdown. checkpoint timeline {}", timelineid);
+    timeline.checkpoint(CheckpointConfig::Forced)?;
+    //TODO Wait for walredo process to shutdown too
```

Comment:
Shouldn't this hook into the existing `tenant_mgr` shutdown logic

Reply:
It's actually vice versa: the `tenant_mgt` calls this `shutdown_all_tenants` that gets here eventually, so not sure this is a good idea to change right now.
Originally it came with https://github.com/zenithdb/zenith/pull/671, I've moved it around and did not dare to change.

Reply:
Ahh I see, my bad